### ISSUE TITLE
Exporter-related metrices support

### ISSUE THREAD

**Issue by mjf** [google/mtail#90]
Hello,

just to notice - there were at least two common metrices suggested to exist for exporters by Mr. Brazil on PromCon 2016: ``<exporter>_up`` and ``<exporter>_scrape_duration_seconds``.

The former one may be helpful as a hint for users to know if the exporter somehow failed not exporting all metrices but keeps running and still exposing some useful metrices (users check it for 0 or 1).

The latter one is pretty self-describing and would be nice to have it as well for ``mtail``.

**Comment by jaqx0r** [comment]
Sounds like a good idea.  Is there a link to the talk video?

**Comment by mjf** [comment]
@jaqx0r Here you are: "[So You Want to Write an Exporter](https://www.youtube.com/watch?v=KXq5ibSj2qA)".

Good resource on various topics concerning writing an exporter is the *Prometheus*' documentation on [writing exporters](https://prometheus.io/docs/instrumenting/writing_exporters/), to be precise the [`<exporter>_scrape_duration_seconds`](https://prometheus.io/docs/instrumenting/writing_exporters/#metrics-about-the-scrape-itself) metric (it would be `mtail_scrape_duration_seconds` in our case) and the paragraph on [failed scrapes](https://prometheus.io/docs/instrumenting/writing_exporters/#failed-scrapes). It seems to me that the documentation covers all the important and interesting things concerning the topic.

**Comment by jaqx0r** [comment]
Thanks!

**Comment by jaqx0r** [comment]
I don't know what `mtail_scrape_duration_seconds` is supposed to measure.  mtail doesn't know how long Prometheus is doing a scrape, Prometheus does tho!  And the docs say that if you don't explicitly send an mtail-up, Prometheus synthesises it based on 2xx vs 5xx HTTP status codes.

I don't think either of these metrics makes sense for mtail.  I can imagine a case where an mtail program fails for a single log line but mtail doesn't keep a track of any of that, and if mtail itself crashes then there's nothing to export.  It sounds like both metrics are designed for systems where an exporter is a proxy for a live system, and mtail does not work like that.  Really mtail has no idea what it is proxying for except the names of the programs it's executing.

The closest match I can imagine is to export a `<program>_up` that is always 1 for every loaded mtail program.  That seems like a waste of effort and bandwidth, though.

**Comment by mjf** [comment]
@jaqx0r Well, I would still suggest to export at least `mtail_up` always beying 1 (in case you want to build some "dynamic" dashboards or whatever weird somebody may need it for).

Your idea with `<program>_up` sounds nicely too but not only it may be waste of effort but also will need some acceptable program name in the allowed metric name syntax (which would be quite complicated to achieve and would lead to possibly ugly, unintended or impractical results too).

The `mtail_scrape_duration_seconds` could have measured the time between `mtail` starts and finishes to export metrices (the time `mtail` is beying scraped). Obviously I was wrong with what the metric should mean too. My idea was to use this metric to reference something like "`mtail`'s performance" and maybe detect some problems with it. Maybe such metric still can be exported using just different name?

So, if neither the `mtail_up` nor the `mtail_scrape_duration_seconds` (or other) makes no sense to you, just close this feature request, please. Thank you.

**Comment by jaqx0r** [comment]
I appreciate trying to make the measures fit mtail.  The best way to measure mtail being scraped is in the /metrics handler itself (which would contain the result of mtail_scrape_duration_seconds) and is going to only measure the time waiting on the metric store lock, and then walking the store to emit each metric.

Elsewhere in https://prometheus.io/docs/instrumenting/writing_exporters/ it says Prometheus collects its own scrape_duration_seconds which would be exactly the same value as the above (plus network time and TCP handshake time, which I choose to ignore for this comparison because the usefulness of this metric, like most metrics, is not in its absolute value but its rate of change.)

Adding `mtail_up` is easy, but I think prometheus' synthetic `up` is equivalent: https://prometheus.io/docs/concepts/jobs_instances/

If we want performance metrics instead, I'm happy to measure VM cycle time and export a histogram of performance per line instead.  But that won't measure what the exporters' document wants.

**Comment by jaqx0r** [comment]
Technically not fixed yet, but I need to refactor all the internal metrics export to support a histogram type before I can implement, so it's just marked as a TODO for now.

**Comment by mjf** [comment]
@jaqx0r OK
